### ISSUE TITLE
Calculator doesn't calculate

### ISSUE THREAD

**Issue by joaquinvacas** [cosmic-utils/calculator#9]
I think I don't have to explain further → _A picture is worth a thousand words_

![image](https://github.com/user-attachments/assets/4be595f4-fd8b-446c-8126-b68224192ce0)

**Comment by joaquinvacas** [comment]
Forgot to mention: Release is the latest from Flathub, 0.1.0

**Comment by edfloreshz** [comment]
Thanks for opening this issue, I'll try to fix this in the next couple days.

**Comment by Joomsy** [comment]
> Forgot to mention: Release is the latest from Flathub, 0.1.0

Same issue when built from source. I had to double take after doing some subtraction, because the result was far from being correct. I'd also get something random each time when trying to do addition.

**Comment by edfloreshz** [comment]
This should be fixed by 046bec08b2bc3179cfc554ec8b692b98dc768f6c, I'll await for feedback before closing this.

**Comment by Joomsy** [comment]
> This should be fixed by [046bec0](https://github.com/cosmic-utils/calculator/commit/046bec08b2bc3179cfc554ec8b692b98dc768f6c), I'll await for feedback before closing this.

Sorry for the late response. I work weekends, so I wasn't able to get back to this until now. Something still seems a little off. At first, addition gave me something wildly incorrect (8+8=66), but after plugging in more things, it seems to have corrected itself. I can't get it to reproduce an incorrect output. Algebraic expressions, however, do give the wrong output. This is actually supposed to be 30.

![screenshot-2024-10-14-22-50-01](https://github.com/user-attachments/assets/442acbb7-c00a-4031-8372-62c984bd10a4)

**Comment by l-const** [comment]
Yeah, the **calculator-rs** dependency doesn't handle operator precedent without parenthesis so it's AST is kind of wrong, if you want to make it work correctly and in an easy way just copy the implementation of the parser & evaluator from here: https://github.com/balajisivaraman/basic_calculator_rs/blob/master/src/main.rs  it uses the **nom** parser library : https://github.com/rust-bakery/nom , famous for those kind of things, there is a book here: https://tfpk.github.io/nominomicon/introduction.html

**Comment by uciharis** [comment]
![screenshot-2024-12-31-11-54-55](https://github.com/user-attachments/assets/23c1c08b-7999-4054-8785-505af6acbc36)

bro, i dont have any idea what i see

**Comment by l-const** [comment]
Have some fixes (there may be other bugs) in an open PR here: https://github.com/0byteme/calculator/pull/1/commits or on  my main branch on my fork: https://github.com/l-const/calculator cc @edfloreshz

**Comment by mmstick** [comment]
calculator-rs doesn't seem to be actively developed or maintained. Perhaps it'd be better to use [evalexpr](https://crates.io/crates/evalexpr) instead. It's still being actively developed.

**Comment by edfloreshz** [comment]
Thanks, I've been busy with other projects but I'll try to make some time for calculator next weekend.

**Comment by edfloreshz** [comment]
Should be solved, we can reopen if we find issues again.

**Comment by l-const** [comment]
did you check integer division @edfloreshz , i am not sure it is handling appropriately from this comment: https://github.com/ISibboI/evalexpr/issues/125#issuecomment-1520257131 they are using a programming evaluation context and not one for gui calculation one? what 4 / 5  outputs?
